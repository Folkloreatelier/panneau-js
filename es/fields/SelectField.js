import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _toConsumableArray from 'babel-runtime/helpers/toConsumableArray';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import isObject from 'lodash/isObject';
import Select, { Async, AsyncCreatable, Creatable } from 'react-select';

import FormGroup from '../FormGroup';

var valuePropTypes = PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array, PropTypes.object]);

var propTypes = {
    name: PropTypes.string,
    label: PropTypes.string,
    value: valuePropTypes,
    options: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.shape({
        value: PropTypes.any,
        label: PropTypes.string
    })])),
    getValueFromOption: PropTypes.func,
    onChange: PropTypes.func,
    onOptionsChange: PropTypes.func,

    placeholder: PropTypes.string,
    noResultsText: PropTypes.string,
    canBeEmpty: PropTypes.bool,
    addEmptyOption: PropTypes.bool,
    emptyOption: PropTypes.shape({
        value: valuePropTypes,
        label: PropTypes.string
    }),
    loadOptions: PropTypes.func,
    fetchOptions: PropTypes.func,
    async: PropTypes.bool,
    multiple: PropTypes.bool,
    searchable: PropTypes.bool,
    clearable: PropTypes.bool,
    creatable: PropTypes.bool,
    style: PropTypes.object // eslint-disable-line
};

var defaultProps = {
    name: null,
    label: '',
    value: null,
    options: [],
    onChange: null,
    onOptionsChange: null,

    getValueFromOption: null,
    canBeEmpty: true,
    searchable: true,
    clearable: true,
    creatable: false,
    async: false,
    multiple: false,
    loadOptions: null,
    fetchOptions: null,
    placeholder: 'Aucun',
    noResultsText: 'Aucun rÃ©sultat',
    addEmptyOption: false,
    emptyOption: { value: '', label: '--' },
    style: null
};

/**
 *  Class: SelectField
 *  Makes a select using react-select
 *
 *  Options are using the format [{ label: "", value: "" }]
 *
 *  Note: value is array if multiple select is allowed
 *  @param {string,number,array,object} value
 *  @return {string,array} newValue
 */

var SelectField = function (_Component) {
    _inherits(SelectField, _Component);

    function SelectField(props) {
        _classCallCheck(this, SelectField);

        var _this = _possibleConstructorReturn(this, (SelectField.__proto__ || Object.getPrototypeOf(SelectField)).call(this, props));

        _this.onChange = _this.onChange.bind(_this);
        _this.onNewOptionClick = _this.onNewOptionClick.bind(_this);
        _this.getValueFromOption = _this.getValueFromOption.bind(_this);

        _this.state = {
            options: props.options
        };
        return _this;
    }

    _createClass(SelectField, [{
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
            var optionsChanged = this.props.options !== nextProps.options;
            if (optionsChanged) {
                this.setState({
                    options: nextProps.options
                });
            }
        }
    }, {
        key: 'onNewOptionClick',
        value: function onNewOptionClick(option) {
            var _this2 = this;

            var options = this.state.options;

            this.setState({
                options: [].concat(_toConsumableArray(options), [option])
            }, function () {
                if (_this2.props.onOptionsChange) {
                    _this2.props.onOptionsChange(_this2.state.options);
                }
            });
        }
    }, {
        key: 'onChange',
        value: function onChange(value) {
            var _props = this.props,
                multiple = _props.multiple,
                clearable = _props.clearable;

            var newValue = void 0;
            if (multiple && value === null) {
                newValue = null;
            } else if (multiple && value.length === 0 && clearable) {
                newValue = null;
            } else if (multiple) {
                newValue = value.map(this.getValueFromOption);
            } else {
                newValue = this.getValueFromOption(value);
            }
            if (this.props.onChange) {
                this.props.onChange(newValue);
            }
        }
    }, {
        key: 'getValueFromOption',
        value: function getValueFromOption(opt) {
            var getValueFromOption = this.props.getValueFromOption;

            if (getValueFromOption !== null) {
                return getValueFromOption(opt);
            }
            return isObject(opt) && typeof opt.value !== 'undefined' ? opt.value : opt;
        }
    }, {
        key: 'loadOptions',
        value: function loadOptions(input, callback) {
            var _props2 = this.props,
                loadOptions = _props2.loadOptions,
                fetchOptions = _props2.fetchOptions;

            if (loadOptions !== null) {
                return loadOptions(input, callback);
            } else if (fetchOptions !== null) {
                return fetchOptions(input).then(function (options) {
                    return {
                        options: options
                    };
                });
            }
            return null;
        }
    }, {
        key: 'render',
        value: function render() {
            var _props3 = this.props,
                name = _props3.name,
                label = _props3.label,
                value = _props3.value,
                canBeEmpty = _props3.canBeEmpty,
                addEmptyOption = _props3.addEmptyOption,
                emptyOption = _props3.emptyOption,
                style = _props3.style,
                async = _props3.async,
                creatable = _props3.creatable,
                multiple = _props3.multiple,
                clearable = _props3.clearable,
                other = _objectWithoutProperties(_props3, ['name', 'label', 'value', 'canBeEmpty', 'addEmptyOption', 'emptyOption', 'style', 'async', 'creatable', 'multiple', 'clearable']);

            var options = this.state.options;


            var selectOptions = [].concat(options);
            if (canBeEmpty && addEmptyOption) {
                selectOptions.unshift(emptyOption);
            }

            var shouldTakeFirstValue = !canBeEmpty && value === null && selectOptions.length > 0;
            var selectValue = shouldTakeFirstValue ? this.getValueFromOption(selectOptions[0]) : value;
            var selectClearable = clearable && shouldTakeFirstValue ? false : clearable;

            var asyncProps = async ? {
                loadOptions: this.loadOptions
            } : {
                options: selectOptions
            };

            var SelectComponent = void 0;
            if (async && creatable) {
                SelectComponent = AsyncCreatable;
            } else if (async) {
                SelectComponent = Async;
            } else if (creatable) {
                SelectComponent = Creatable;
            } else {
                SelectComponent = Select;
            }

            return React.createElement(
                FormGroup,
                _extends({}, other, {
                    className: 'form-group-select',
                    name: name,
                    label: label
                }),
                React.createElement(
                    'div',
                    { style: style },
                    React.createElement(SelectComponent, _extends({
                        name: 'form-field-select',
                        multi: multiple
                    }, other, asyncProps, {
                        clearable: selectClearable,
                        value: selectValue,
                        onChange: this.onChange,
                        onNewOptionClick: this.onNewOptionClick
                    }))
                )
            );
        }
    }]);

    return SelectField;
}(Component);

SelectField.propTypes = propTypes;
SelectField.defaultProps = defaultProps;

export default SelectField;