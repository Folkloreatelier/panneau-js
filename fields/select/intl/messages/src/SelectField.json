[
  {
    "id": "fields.select.placeholder",
    "defaultMessage": "Select..."
  },
  {
    "id": "fields.select.no_options",
    "defaultMessage": "No options"
  },
  {
    "id": "fields.select.loading",
    "defaultMessage": "Loading..."
  },
  {
    "id": "fields.select.screen_reader_status",
    "defaultMessage": "{count, plural, =0 {no result} one {# result} other {# results}} available"
  }
]