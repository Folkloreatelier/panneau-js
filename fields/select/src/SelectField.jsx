/* eslint-disable react/no-array-index-key, react/button-has-type, react/jsx-props-no-spreading */
import React, { useMemo, useCallback } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Select from 'react-select';
import isEqual from 'lodash/isEqual';
import { PropTypes as PanneauPropTypes } from '@panneau/core';
import { getSelectOptions } from '@panneau/core/utils';

import styles from './styles.module.scss';

const propTypes = {
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
    options: PanneauPropTypes.selectOptions,
    disabled: PropTypes.bool,
    isMulti: PropTypes.bool,
    isSearchable: PropTypes.bool,
    withoutReset: PropTypes.bool,
    noOptionsMessage: PropTypes.string,
    placeholder: PropTypes.string,
    className: PropTypes.string,
    onChange: PropTypes.func,
};

const defaultProps = {
    value: null,
    options: [],
    disabled: false,
    isMulti: false,    
    isSearchable: false,
    withoutReset: false,
    noOptionsMessage: 'No results.',
    placeholder: 'Select...',
    className: null,
    onChange: null,
};

const SelectField = ({
    value,
    options,
    disabled,
    isMulti,
    isSearchable,
    withoutReset,
    noOptionsMessage,
    placeholder,
    className,
    onChange,
    ...props
}) => {
    const finalOptions = useMemo(() => getSelectOptions(options), [options]);
    const onChangeOption = useCallback(
        (newValue) => {
            if (onChange !== null) {
                if (isMulti) {
                    onChange(newValue !== null ? newValue.map(({ value: val }) => val) : null);
                } else {
                    onChange(newValue !== null && newValue.value ? newValue.value : null);
                }
            }
        },
        [onChange, isMulti],
    );
    const optionValue = useMemo(() => {
        if (isMulti) {
            return value !== null
                ? value.map((val) =>
                      finalOptions.find((opt) =>
                          opt.value !== null ? isEqual(val, opt.value) : false,
                      ),
                  )
                : [];
        }
        return finalOptions.find((opt) => (opt.value !== null ? isEqual(value, opt.value) : false));
    }, [value, options, isMulti]);

    return (
        <Select
            className={classNames([
                styles.container,
                {
                    [className]: className !== null,
                },
            ])}
            {...props}
            value={optionValue || null}
            options={finalOptions}
            disabled={disabled}
            isMulti={isMulti}
            isClearable={!withoutReset}
            isSearchable={isSearchable}
            noOptionsMessage={() => noOptionsMessage}
            placeholder={placeholder}
            onChange={onChangeOption}
        />
    );
};

SelectField.propTypes = propTypes;
SelectField.defaultProps = defaultProps;

export default SelectField;
