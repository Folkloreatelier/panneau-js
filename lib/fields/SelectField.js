'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _isObject = require('lodash/isObject');

var _isObject2 = _interopRequireDefault(_isObject);

var _reactSelect = require('react-select');

var _reactSelect2 = _interopRequireDefault(_reactSelect);

var _FormGroup = require('../FormGroup');

var _FormGroup2 = _interopRequireDefault(_FormGroup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var valuePropTypes = _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number, _propTypes2.default.array, _propTypes2.default.object]);

var propTypes = {
    name: _propTypes2.default.string,
    label: _propTypes2.default.string,
    value: valuePropTypes,
    options: _propTypes2.default.arrayOf(_propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.shape({
        value: _propTypes2.default.any,
        label: _propTypes2.default.string
    })])),
    getValueFromOption: _propTypes2.default.func,
    onChange: _propTypes2.default.func,
    onOptionsChange: _propTypes2.default.func,

    placeholder: _propTypes2.default.string,
    noResultsText: _propTypes2.default.string,
    canBeEmpty: _propTypes2.default.bool,
    addEmptyOption: _propTypes2.default.bool,
    emptyOption: _propTypes2.default.shape({
        value: valuePropTypes,
        label: _propTypes2.default.string
    }),
    loadOptions: _propTypes2.default.func,
    fetchOptions: _propTypes2.default.func,
    async: _propTypes2.default.bool,
    multiple: _propTypes2.default.bool,
    searchable: _propTypes2.default.bool,
    clearable: _propTypes2.default.bool,
    creatable: _propTypes2.default.bool,
    style: _propTypes2.default.object // eslint-disable-line
};

var defaultProps = {
    name: null,
    label: '',
    value: null,
    options: [],
    onChange: null,
    onOptionsChange: null,

    getValueFromOption: null,
    canBeEmpty: true,
    searchable: true,
    clearable: true,
    creatable: false,
    async: false,
    multiple: false,
    loadOptions: null,
    fetchOptions: null,
    placeholder: 'Aucun',
    noResultsText: 'Aucun rÃ©sultat',
    addEmptyOption: false,
    emptyOption: { value: '', label: '--' },
    style: null
};

/**
 *  Class: SelectField
 *  Makes a select using react-select
 *
 *  Options are using the format [{ label: "", value: "" }]
 *
 *  Note: value is array if multiple select is allowed
 *  @param {string,number,array,object} value
 *  @return {string,array} newValue
 */

var SelectField = function (_Component) {
    (0, _inherits3.default)(SelectField, _Component);

    function SelectField(props) {
        (0, _classCallCheck3.default)(this, SelectField);

        var _this = (0, _possibleConstructorReturn3.default)(this, (SelectField.__proto__ || Object.getPrototypeOf(SelectField)).call(this, props));

        _this.onChange = _this.onChange.bind(_this);
        _this.onNewOptionClick = _this.onNewOptionClick.bind(_this);
        _this.getValueFromOption = _this.getValueFromOption.bind(_this);

        _this.state = {
            options: props.options
        };
        return _this;
    }

    (0, _createClass3.default)(SelectField, [{
        key: 'componentWillReceiveProps',
        value: function componentWillReceiveProps(nextProps) {
            var optionsChanged = this.props.options !== nextProps.options;
            if (optionsChanged) {
                this.setState({
                    options: nextProps.options
                });
            }
        }
    }, {
        key: 'onNewOptionClick',
        value: function onNewOptionClick(option) {
            var _this2 = this;

            var options = this.state.options;

            this.setState({
                options: [].concat((0, _toConsumableArray3.default)(options), [option])
            }, function () {
                if (_this2.props.onOptionsChange) {
                    _this2.props.onOptionsChange(_this2.state.options);
                }
            });
        }
    }, {
        key: 'onChange',
        value: function onChange(value) {
            var _props = this.props,
                multiple = _props.multiple,
                clearable = _props.clearable;

            var newValue = void 0;
            if (multiple && value === null) {
                newValue = null;
            } else if (multiple && value.length === 0 && clearable) {
                newValue = null;
            } else if (multiple) {
                newValue = value.map(this.getValueFromOption);
            } else {
                newValue = this.getValueFromOption(value);
            }
            if (this.props.onChange) {
                this.props.onChange(newValue);
            }
        }
    }, {
        key: 'getValueFromOption',
        value: function getValueFromOption(opt) {
            var getValueFromOption = this.props.getValueFromOption;

            if (getValueFromOption !== null) {
                return getValueFromOption(opt);
            }
            return (0, _isObject2.default)(opt) && typeof opt.value !== 'undefined' ? opt.value : opt;
        }
    }, {
        key: 'loadOptions',
        value: function loadOptions(input, callback) {
            var _props2 = this.props,
                loadOptions = _props2.loadOptions,
                fetchOptions = _props2.fetchOptions;

            if (loadOptions !== null) {
                return loadOptions(input, callback);
            } else if (fetchOptions !== null) {
                return fetchOptions(input).then(function (options) {
                    return {
                        options: options
                    };
                });
            }
            return null;
        }
    }, {
        key: 'render',
        value: function render() {
            var _props3 = this.props,
                name = _props3.name,
                label = _props3.label,
                value = _props3.value,
                canBeEmpty = _props3.canBeEmpty,
                addEmptyOption = _props3.addEmptyOption,
                emptyOption = _props3.emptyOption,
                style = _props3.style,
                async = _props3.async,
                creatable = _props3.creatable,
                multiple = _props3.multiple,
                clearable = _props3.clearable,
                other = (0, _objectWithoutProperties3.default)(_props3, ['name', 'label', 'value', 'canBeEmpty', 'addEmptyOption', 'emptyOption', 'style', 'async', 'creatable', 'multiple', 'clearable']);
            var options = this.state.options;


            var selectOptions = [].concat(options);
            if (canBeEmpty && addEmptyOption) {
                selectOptions.unshift(emptyOption);
            }

            var shouldTakeFirstValue = !canBeEmpty && value === null && selectOptions.length > 0;
            var selectValue = shouldTakeFirstValue ? this.getValueFromOption(selectOptions[0]) : value;
            var selectClearable = clearable && shouldTakeFirstValue ? false : clearable;

            var asyncProps = async ? {
                loadOptions: this.loadOptions
            } : {
                options: selectOptions
            };

            var SelectComponent = void 0;
            if (async && creatable) {
                SelectComponent = _reactSelect.AsyncCreatable;
            } else if (async) {
                SelectComponent = _reactSelect.Async;
            } else if (creatable) {
                SelectComponent = _reactSelect.Creatable;
            } else {
                SelectComponent = _reactSelect2.default;
            }

            return _react2.default.createElement(
                _FormGroup2.default,
                (0, _extends3.default)({}, other, {
                    className: 'form-group-select',
                    name: name,
                    label: label
                }),
                _react2.default.createElement(
                    'div',
                    { style: style },
                    _react2.default.createElement(SelectComponent, (0, _extends3.default)({
                        name: 'form-field-select',
                        multi: multiple
                    }, other, asyncProps, {
                        clearable: selectClearable,
                        value: selectValue,
                        onChange: this.onChange,
                        onNewOptionClick: this.onNewOptionClick
                    }))
                )
            );
        }
    }]);
    return SelectField;
}(_react.Component);

SelectField.propTypes = propTypes;
SelectField.defaultProps = defaultProps;

exports.default = SelectField;