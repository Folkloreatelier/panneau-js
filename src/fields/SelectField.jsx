import React, { Component } from 'react';
import PropTypes from 'prop-types';
import isObject from 'lodash/isObject';
import Select, { Async, AsyncCreatable, Creatable } from 'react-select';
import '../styles/vendor.global.scss';
import FormGroup from '../FormGroup';

const valuePropTypes = PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.array,
    PropTypes.object,
]);

const propTypes = {
    name: PropTypes.string,
    label: PropTypes.string,
    value: valuePropTypes,
    options: PropTypes.arrayOf(PropTypes.oneOf([
        PropTypes.string,
        PropTypes.shape({
            value: PropTypes.any,
            label: PropTypes.string,
        }),
    ])),
    getValueFromOption: PropTypes.func,
    onChange: PropTypes.func,
    onOptionsChange: PropTypes.func,

    placeholder: PropTypes.string,
    noResultsText: PropTypes.string,
    canBeEmpty: PropTypes.bool,
    addEmptyOption: PropTypes.bool,
    emptyOption: PropTypes.shape({
        value: valuePropTypes,
        label: PropTypes.string,
    }),
    loadOptions: PropTypes.func,
    fetchOptions: PropTypes.func,
    async: PropTypes.bool,
    multiple: PropTypes.bool,
    searchable: PropTypes.bool,
    clearable: PropTypes.bool,
    creatable: PropTypes.bool,
    style: PropTypes.object, // eslint-disable-line
};

const defaultProps = {
    name: null,
    label: '',
    value: null,
    options: [],
    onChange: null,
    onOptionsChange: null,

    getValueFromOption: null,
    canBeEmpty: true,
    searchable: true,
    clearable: true,
    creatable: false,
    async: false,
    multiple: false,
    loadOptions: null,
    fetchOptions: null,
    placeholder: 'Aucun',
    noResultsText: 'Aucun rÃ©sultat',
    addEmptyOption: false,
    emptyOption: { value: '', label: '--' },
    style: null,
};

/**
 *  Class: SelectField
 *  Makes a select using react-select
 *
 *  Options are using the format [{ label: "", value: "" }]
 *
 *  Note: value is array if multiple select is allowed
 *  @param {string,number,array,object} value
 *  @return {string,array} newValue
 */

class SelectField extends Component {
    constructor(props) {
        super(props);
        this.onChange = this.onChange.bind(this);
        this.onNewOptionClick = this.onNewOptionClick.bind(this);
        this.getValueFromOption = this.getValueFromOption.bind(this);

        this.state = {
            options: props.options,
        };
    }

    componentWillReceiveProps(nextProps) {
        const optionsChanged = this.props.options !== nextProps.options;
        if (optionsChanged) {
            this.setState({
                options: nextProps.options,
            });
        }
    }

    onNewOptionClick(option) {
        const { options } = this.state;
        this.setState({
            options: [
                ...options,
                option,
            ],
        }, () => {
            if (this.props.onOptionsChange) {
                this.props.onOptionsChange(this.state.options);
            }
        });
    }

    onChange(value) {
        const { multiple, clearable } = this.props;
        let newValue;
        if (multiple && value === null) {
            newValue = null;
        } else if (multiple && value.length === 0 && clearable) {
            newValue = null;
        } else if (multiple) {
            newValue = value.map(this.getValueFromOption);
        } else {
            newValue = this.getValueFromOption(value);
        }
        if (this.props.onChange) {
            this.props.onChange(newValue);
        }
    }

    getValueFromOption(opt) {
        const { getValueFromOption } = this.props;
        if (getValueFromOption !== null) {
            return getValueFromOption(opt);
        }
        return isObject(opt) && typeof opt.value !== 'undefined' ? opt.value : opt;
    }

    loadOptions(input, callback) {
        const { loadOptions, fetchOptions } = this.props;
        if (loadOptions !== null) {
            return loadOptions(input, callback);
        } else if (fetchOptions !== null) {
            return fetchOptions(input).then(options => ({
                options,
            }));
        }
        return null;
    }

    render() {
        const {
            name,
            label,
            value,
            canBeEmpty,
            addEmptyOption,
            emptyOption,
            style,
            async,
            creatable,
            multiple,
            clearable,
            ...other
        } = this.props;

        const {
            options,
        } = this.state;

        const selectOptions = [].concat(options);
        if (canBeEmpty && addEmptyOption) {
            selectOptions.unshift(emptyOption);
        }

        const shouldTakeFirstValue = !canBeEmpty && value === null && selectOptions.length > 0;
        const selectValue = shouldTakeFirstValue ?
            this.getValueFromOption(selectOptions[0]) : value;
        const selectClearable = clearable && shouldTakeFirstValue ? false : clearable;

        const asyncProps = async ? {
            loadOptions: this.loadOptions,
        } : {
            options: selectOptions,
        };

        let SelectComponent;
        if (async && creatable) {
            SelectComponent = AsyncCreatable;
        } else if (async) {
            SelectComponent = Async;
        } else if (creatable) {
            SelectComponent = Creatable;
        } else {
            SelectComponent = Select;
        }

        return (
            <FormGroup
                {...other}
                className="form-group-select"
                name={name}
                label={label}
            >
                <div style={style}>
                    <SelectComponent
                        name="form-field-select"
                        multi={multiple}
                        {...other}
                        {...asyncProps}
                        clearable={selectClearable}
                        value={selectValue}
                        onChange={this.onChange}
                        onNewOptionClick={this.onNewOptionClick}
                    />
                </div>
            </FormGroup>
        );
    }
}

SelectField.propTypes = propTypes;
SelectField.defaultProps = defaultProps;

export default SelectField;
