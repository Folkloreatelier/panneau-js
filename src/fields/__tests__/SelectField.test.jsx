import React from 'react';
import renderer from 'react-test-renderer';
import SelectField from '../SelectField';

test('SelectField render a <div />', () => {
    const component = renderer.create(
        <SelectField />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
