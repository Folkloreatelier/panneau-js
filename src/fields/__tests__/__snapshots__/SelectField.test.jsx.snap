// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField render a <div /> 1`] = `
<div
  className="form-group form-group-select"
>
  <div
    className="form-group-inner"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <div
      style={null}
    >
      <div
        className="Select Select--single is-clearable is-searchable"
        style={undefined}
      >
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={undefined}
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-2--value"
          >
            <div
              className="Select-placeholder"
            >
              Aucun
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-activedescendant="react-select-2--value"
                aria-describedby={undefined}
                aria-expanded="false"
                aria-haspopup="false"
                aria-label={undefined}
                aria-labelledby={undefined}
                aria-owns=""
                className={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                tabIndex={undefined}
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
